### PR TITLE
fix: recursive call makes duplicate entries in registrar

### PR DESCRIPTION
**Pull request by Zoly**
When assigning new class to registrar the class initialization makes a recursive call.

By setting the didDiscovery variable to true before the assignment will prevent entry duplication in the registrar.

**Description**
Explain what you have changed, and why.

**Checklist:**
- [ ] Securely signed commits
- [ ] Component(s) with PHPDoc blocks, only if necessary or adds value
- [ ] Unit testing, with >80% coverage
- [ ] User guide updated
- [ ] Conforms to style guide

### CODE REVIEW

**Comment by kenjis** [comment]
Thank you for sending this PR!

We expect the following in all Pull Requests (PRs).
- PRs must be sent to the [appropriate branch](https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/pull_request.md#branching)
- All git commits must be [GPG-signed](https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/pull_request.md#signing)
- Must follow our [style guide](https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/pull_request.md#php-style)
- Be [commented](https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/pull_request.md#comments) in the PHP source file
- Be documented in the [user guide](https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/pull_request.md#user-guide)
- Be [unit tested](https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/pull_request.md#unit-testing)
- Pass all checks in GitHub Actions

> [!IMPORTANT]
> We expect all code changes or bug-fixes to be accompanied by one or more tests added to our test suite to prove the code works.

If pull requests do not comply with the above, they will likely be closed. Since we are a team of volunteers, we don't have any more time to work on the framework than you do. Please make it as painless for your contributions to be included as possible.

See https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/pull_request.md

**Comment by Zoly** [comment]
Sadly this patch is not a good solution!

After digging trough the code for two days I was so overjoyed to find the origin point of the issue with the debug toolbar duplicated elements that I became to eager with the solution I found.

By removing the seemingly duplicate entries in the registrar do indeed resolves the toolbar duplicate items problem, but leads to rendering issues.

I compared the duplicated items and they are identical. That leaves us with the conclusion that the rendering process uses the registrar as a map for in which order and how many times to execute the rendering process. Meanwhile the toolbar thinks of the registrar as a unique list of registered packages. The two usage modes conflict with each other and if one is solved the other will fail.

This is an issue that needs further deliberation and clarification in the usage scope.

**Comment by Zoly** [comment]
I just discovered that the debug toolbar doesn't use the registrar to generate the toolbar. It uses the config that is generated based on the registrar, in the same function, right after it finished populating the registrar. 

After restoring the original from repo, I made a single change to use unique list to populate the config properties:
```
        // Check the registrar class for a method named after this class' shortName
        foreach (array_unique(static::$registrars, SORT_REGULAR) as $callable) {
```
and the rendering is functional with the toolbar duplicate elements issue also resolved, but I'm not sure if there are no other surprises with this change even if it makes sense (the fist one did to at the time).

**Comment by Zoly** [comment]
This patch may make the registrar unique, but will affect the rendering process.

This all started with the fact that on my end the debug toolbar has duplicate items for more then on composer package I installed. After digging for two days in the code I arrived at the source of the problem, the during the registration process the newly associated classes make recursive calls and ends up registering the packages multiple times in the registrar, thus when the debug toolbar is generate it's elements the toolbar items are duplicated. So this patch came to be, but later I found some buttons incorrectly rendered on same pages, that are previously worked. So I concluded that the two processes expect different things from the same registrar. 

During the testing process I discovered that the duplicated registrar leads to duplicated config properties, but only in case the property is an index based array. To merge the data array_merge() is used, with behaves differently for indexed and associative array respectively:

> If the input arrays have the same string keys, then the later value for that key will overwrite the previous one. If, however, the arrays contain numeric keys, the later value will not overwrite the original value, but will be appended. Values in the input array with numeric keys will be renumbered with incrementing keys starting from zero in the result array.

The registrar is duplicated, and that leads to some of the config values to be duplicated as well, including the collectors property used by the debug toolbar.

When I tried to make the config property generation unique, the rendering process broke. At this point I've given up on fixing the toolbar problem from the roots and just did a dirty patch in the toolbar service making the collectors property there unique with good results.

In conclusion, the registrar process needs to be revisited by brighter minds then mine, because there are two implementations that are expecting different things from the same property. The rendering process uses the duplicated items as a map to how many times and in witch order to apply the rendering, meanwhile the toolbar expects to see a unique list of registered packages to be processed.
